src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -235,7 +238,7 @@ private void ReadDirectoryChangesCallback(uint errorCode, uint numBytes, AsyncRe
                 }
                 else
                 {
-                    ParseEventBufferAndNotifyForEach(state.Buffer);
+                    ParseEventBufferAndNotifyForEach(new ReadOnlySpan<byte>(state.Buffer, 0, (int)numBytes));
```

Comment:
Unlike #42474, this will throw when the OS returns invalid size. It is impossible to reason about the state in this case and the state of the buffer. I think it is better to throw in this case instead of trimming it to the expected buffer size.

Reply:
I am not sure if it is better to throw. If proven we can get bogus values inside the array then we may expect get wrong size too. 

I would prefer to check and just exit. Diagnostics would be difficult especially in term getting a reliable repro for the issue.

---
In reply to: [491177407](https://github.com/dotnet/runtime/pull/42474#discussion_r491177407) [](ancestors = 491177407)

Reply:
If the OS returns bogus buffer size, there is a high chance that something went horribly wrong and the buffer is filled with uninitialized data or there is a bad buffer overrun. Letting this method work on buffer with uninitialized data is the worst thing you can imagine.

> Diagnostics would be difficult especially in term getting a reliable repro for the issue.

You will be at least able to get a crash dump that captures the state of the process. If you ignore the problem, you will likely see some kind of corrupted data later and will have much harder time tracking it back to the root cause.

Reply:
Getting crash dump is not always happening. This issue reported on github without any dump and no clue. We were lucky to get a dump later from a real customer. Anyway, this is minore issue shouldn't block this change.

Reply:
> This issue reported on github without any dump and no clue. 

I do not think it was the case. From https://github.com/dotnet/runtime/issues/40412#issue-673817137: `Dump File:	memory.hdmp : C:\tmp\dumps\Dump1\memory.hdmp`

Reply:
There was dump which we couldn't get https://github.com/dotnet/runtime/issues/40412#issuecomment-671620810

Reply:
I would not describe that as no dump and no clue. We had a dump and the dump provided a clue where the problem is. We just were not able to get our hands on the dump ourselves to see the details due to privacy issues. We have a number of ways how we can make progress in situations like that.

Reply:
fair enough.